pkg/controllers/restore/controller.go

```diff
@@ -700,3 +719,72 @@ func (h *handler) setReconcilingCondition(restore *v1.Restore, originalErr error
 
 	return restore, err
 }
+
+func (h *handler) Lock(restore *v1.Restore) error {
+	lease, err := h.kubernetesLeaseClient.Get(h.ctx, leaseName, k8sv1.GetOptions{})
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			return err
+		}
+		lease = &coordinationv1.Lease{
+			ObjectMeta: k8sv1.ObjectMeta{
+				Name:      leaseName,
+				Namespace: util.ChartNamespace,
+			},
+			Spec: coordinationv1.LeaseSpec{
+				HolderIdentity: leaseHolderName(restore),
+			},
+		}
+		_, err = h.kubernetesLeaseClient.Create(h.ctx, lease, k8sv1.CreateOptions{})
+		return err
+	}
+
+	if lease.Spec.HolderIdentity != nil {
+		// check that the current holder exists and is not deleted from cluster (if Restore CR gets deleted, reset lease)
+		split := strings.SplitN(*lease.Spec.HolderIdentity, ":", 2)
+		if len(split) != 2 {
+			logrus.Errorf("Invalid holderIdentity on lease, resetting to current Restore")
+			return h.updateLeaseHolderIdentity(restore, lease)
+		}
+		restoreName, restoreUID := split[0], split[1]
+		restoreInProgress, err := h.restores.Get(restoreName, k8sv1.GetOptions{})
+		if err == nil {
+			// restore CR by this name exists.
+			if string(restoreInProgress.UID) != restoreUID {
+				// the Restore CR by this name has a different UID, update lease lock with it
+				return h.updateLeaseHolderIdentity(restoreInProgress, lease)
+			}
+		} else if apierrors.IsNotFound(err) {
+			// restore CR no longer exists in cluster, remove it from holderIdentity and set current Restore CR as holder
+			return h.updateLeaseHolderIdentity(restore, lease)
+		}
+		return fmt.Errorf("restore %v is in progress", restoreName)
+	}
+	return h.updateLeaseHolderIdentity(restore, lease)
+}
+
+func (h *handler) updateLeaseHolderIdentity(restore *v1.Restore, lease *coordinationv1.Lease) error {
+	lease.Spec.HolderIdentity = leaseHolderName(restore)
+	_, err := h.kubernetesLeaseClient.Update(h.ctx, lease, k8sv1.UpdateOptions{})
+	return err
+}
+
+func (h *handler) Unlock(id string) error {
+	lease, err := h.kubernetesLeaseClient.Get(h.ctx, leaseName, k8sv1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	if lease.Spec.HolderIdentity == nil {
+		return nil
+	}
+	if *lease.Spec.HolderIdentity != id {
+		return fmt.Errorf("lease holder ID %q does not match existing restore CR", id)
```

Comment:
Can this error be clarified? We are attempting to unlock and the ID does not match between the restore trying to unlock and what exists on the lease